jwt.verify.publickey.location property disable ssl hostname verify?

**i set jwt.verify.publickey.location property in my properties:**

mp.jwt.verify.publickey.location=https://userservice/.well-known/jwks.json
mp.jwt.verify.issuer=TelCloud

**and i'm use quarkus-smallrye-jwt, and it's version is 2.4.1.Final**

**but I've had such a mistake:**

Caused by: javax.net.ssl.SSLHandshakeException: No subject alternative DNS name matching userservice found.
        at java.base/sun.security.ssl.Alert.createSSLException(Unknown Source)
        at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
        at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
        at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
        at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(Unknown Source)
        at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(Unknown Source)
        at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(Unknown Source)
        at java.base/sun.security.ssl.SSLHandshake.consume(Unknown Source)
        at java.base/sun.security.ssl.HandshakeContext.dispatch(Unknown Source)
        at java.base/sun.security.ssl.HandshakeContext.dispatch(Unknown Source)
        at java.base/sun.security.ssl.TransportContext.dispatch(Unknown Source)
        at java.base/sun.security.ssl.SSLTransport.decode(Unknown Source)
        at java.base/sun.security.ssl.SSLSocketImpl.decode(Unknown Source)
        at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(Unknown Source)
        at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(Unknown Source)
        at java.base/sun.net.www.protocol.https.HttpsClient.afterConnect(Unknown Source)
        at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(Unknown Source)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(Unknown Source)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(Unknown Source)
        at java.base/java.net.HttpURLConnection.getResponseCode(Unknown Source)
        at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(Unknown Source)
        at org.jose4j.http.Get.get(Get.java:80)
        at org.jose4j.jwk.HttpsJwks.refresh(HttpsJwks.java:204)
        at io.smallrye.jwt.auth.principal.AbstractKeyLocationResolver.isHttpsJwksInitialized(AbstractKeyLocationResolver.java:100)
        at io.smallrye.jwt.auth.principal.KeyLocationResolver.initializeKeyContent(KeyLocationResolver.java:89)
        at io.smallrye.jwt.auth.principal.KeyLocationResolver.<init>(KeyLocationResolver.java:45)
        ... 81 more
Caused by: java.security.cert.CertificateException: No subject alternative DNS name matching userservice found.
        at java.base/sun.security.util.HostnameChecker.matchDNS(Unknown Source)
        at java.base/sun.security.util.HostnameChecker.match(Unknown Source)
        at java.base/sun.security.ssl.X509TrustManagerImpl.checkIdentity(Unknown Source)
        at java.base/sun.security.ssl.X509TrustManagerImpl.checkIdentity(Unknown Source)
        at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(Unknown Source)
        at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(Unknown Source)
        ... 103 more

**so i want to diasble ssl hostname verify, who can help me?**

--- a/doc/modules/ROOT/pages/configuration.adoc
+++ b/doc/modules/ROOT/pages/configuration.adoc
@@ -61,6 +61,9 @@ SmallRye JWT supports many properties which can be used to customize the token p
 |smallrye.jwt.decrypt.key.location|none|Config property allows for an external or internal location of Private Decryption Key to be specified. This property is deprecated, use `mp.jwt.decrypt.key.location`.
 |smallrye.jwt.decrypt.algorithm|`RSA_OAEP`|Decryption algorithm.
 |smallrye.jwt.token.decryption.kid|none|Decryption Key identifier. If it is set then the decryption JWK key as well every JWT token must have a matching `kid` header.
+|smallrye.jwt.tls.certificate.path|none|Path to TLS trusted certificate which may need to be configured if the keys have to be fetched over `HTTPS`.
+|smallrye.jwt.tls.trust-all|false|Trust all the hostnames. If the keys have to be fetched over `HTTPS` and this property is set to `true` then all the hostnames are trusted by default.
+|smallrye.jwt.tls.trusted.hosts|none|Set of trusted hostnames. If the keys have to be fetched over `HTTPS` and `smallrye.jwt.tls.trust-all` is set to `false` then then this property can be used to configure the trusted hostnames.
 |===
 
 = Create JsonWebToken with JWTParser

--- a/implementation/jwt-auth/src/main/java/io/smallrye/jwt/auth/principal/AbstractKeyLocationResolver.java
+++ b/implementation/jwt-auth/src/main/java/io/smallrye/jwt/auth/principal/AbstractKeyLocationResolver.java
@@ -22,7 +22,12 @@ import java.security.Key;
 import java.security.cert.X509Certificate;
 import java.util.Base64;
 import java.util.List;
+import java.util.Set;
 
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLSession;
+
+import org.jose4j.http.Get;
 import org.jose4j.jwk.HttpsJwks;
 import org.jose4j.jwk.JsonWebKey;
 import org.jose4j.jwk.OctetSequenceJsonWebKey;
@@ -39,6 +44,7 @@ import io.smallrye.jwt.util.ResourceUtils.UrlStreamResolver;
  * This implements the MP-JWT 1.1 mp.jwt.verify.publickey.location config property resolution logic
  */
 public class AbstractKeyLocationResolver {
+
     private static final String HTTP_SCHEME = "http:";
     private static final String HTTPS_SCHEME = "https:";
 
@@ -85,14 +91,37 @@ public class AbstractKeyLocationResolver {
         return jws.getHeaders().getStringHeaderValue(JsonWebKey.KEY_ID_PARAMETER);
     }
 
-    protected HttpsJwks initializeHttpsJwks(String location) throws IOException {
+    protected HttpsJwks initializeHttpsJwks(String location)
+            throws IOException {
         PrincipalLogging.log.tryCreateKeyFromHttpsJWKS();
+        HttpsJwks theHttpsJwks = getHttpsJwks(location);
+        if (location.startsWith(HTTPS_SCHEME)) {
+            Get httpGet = getHttpGet();
+            if (authContextInfo.isTlsTrustAll()) {
+                httpGet.setHostnameVerifier(new TrustAllHostnameVerifier());
+            } else if (authContextInfo.getTlsTrustedHosts() != null) {
+                httpGet.setHostnameVerifier(new TrustedHostsHostnameVerifier(authContextInfo.getTlsTrustedHosts()));
+            }
+            if (authContextInfo.getTlsCertificatePath() != null) {
+                httpGet.setTrustedCertificates(loadPEMCertificate(readKeyContent(authContextInfo.getTlsCertificatePath())));
+            }
+            theHttpsJwks.setSimpleHttpGet(httpGet);
+        }
+        return theHttpsJwks;
+    }
+
+    protected HttpsJwks getHttpsJwks(String location) {
         HttpsJwks theHttpsJwks = new HttpsJwks(location);
         theHttpsJwks.setDefaultCacheDuration(authContextInfo.getJwksRefreshInterval().longValue() * 60L);
         return theHttpsJwks;
     }
 
-    protected boolean isHttpsJwksInitialized(String keyLocation) throws IOException {
+    protected Get getHttpGet() {
+        return new Get();
+    }
+
+    protected boolean isHttpsJwksInitialized(String keyLocation)
+            throws IOException {
         if (mayBeFormat(KeyFormat.JWK) && keyLocation != null
                 && (keyLocation.startsWith(HTTPS_SCHEME) || keyLocation.startsWith(HTTP_SCHEME))) {
             httpsJwks = initializeHttpsJwks(keyLocation);
@@ -287,5 +316,28 @@ public class AbstractKeyLocationResolver {
             PrincipalLogging.log.keyContentIsNotValidEncodedPEMCertificate(e);
         }
         return cert;
+    }
+
+    static class TrustAllHostnameVerifier implements HostnameVerifier {
+
+        @Override
+        public boolean verify(String hostname, SSLSession session) {
+            return true;
+        }
+    }
+
+    static class TrustedHostsHostnameVerifier implements HostnameVerifier {
+
+        Set<String> hosts;
+
+        TrustedHostsHostnameVerifier(Set<String> hosts) {
+            this.hosts = hosts;
+        }
+
+        @Override
+        public boolean verify(String hostname, SSLSession session) {
+            return hosts.contains(hostname);
+        }
+
     }
 }

--- a/implementation/jwt-auth/src/main/java/io/smallrye/jwt/auth/principal/JWTAuthContextInfo.java
+++ b/implementation/jwt-auth/src/main/java/io/smallrye/jwt/auth/principal/JWTAuthContextInfo.java
@@ -64,6 +64,9 @@ public class JWTAuthContextInfo {
     private Set<String> requiredClaims;
     private boolean relaxVerificationKeyValidation = true;
     private boolean verifyCertificateThumbprint;
+    private String tlsCertificatePath;
+    private Set<String> tlsTrustedHosts;
+    private boolean tlsTrustAll;
 
     public JWTAuthContextInfo() {
     }
@@ -121,6 +124,9 @@ public class JWTAuthContextInfo {
         this.requiredClaims = orig.getRequiredClaims();
         this.relaxVerificationKeyValidation = orig.isRelaxVerificationKeyValidation();
         this.verifyCertificateThumbprint = orig.isVerifyCertificateThumbprint();
+        this.tlsCertificatePath = orig.getTlsCertificatePath();
+        this.tlsTrustedHosts = orig.getTlsTrustedHosts();
+        this.tlsTrustAll = orig.isTlsTrustAll();
     }
 
     @Deprecated
@@ -423,5 +429,29 @@ public class JWTAuthContextInfo {
 
     public void setVerifyCertificateThumbprint(boolean verifyCertificateThumbprint) {
         this.verifyCertificateThumbprint = verifyCertificateThumbprint;
+    }
+
+    public String getTlsCertificatePath() {
+        return tlsCertificatePath;
+    }
+
+    public void setTlsCertificatePath(String tlsCertificatePath) {
+        this.tlsCertificatePath = tlsCertificatePath;
+    }
+
+    public Set<String> getTlsTrustedHosts() {
+        return tlsTrustedHosts;
+    }
+
+    public void setTlsTrustedHosts(Set<String> tlsTrustedHosts) {
+        this.tlsTrustedHosts = tlsTrustedHosts;
+    }
+
+    public void setTlsTrustAll(boolean tlsTrustAll) {
+        this.tlsTrustAll = tlsTrustAll;
+    }
+
+    public boolean isTlsTrustAll() {
+        return this.tlsTrustAll;
     }
 }

--- a/implementation/jwt-auth/src/main/java/io/smallrye/jwt/config/JWTAuthContextInfoProvider.java
+++ b/implementation/jwt-auth/src/main/java/io/smallrye/jwt/config/JWTAuthContextInfoProvider.java
@@ -128,6 +128,8 @@ public class JWTAuthContextInfoProvider {
         provider.expectedAudience = Optional.empty();
         provider.groupsSeparator = DEFAULT_GROUPS_SEPARATOR;
         provider.requiredClaims = Optional.empty();
+        provider.tlsCertificatePath = Optional.empty();
+        provider.tlsTrustedHosts = Optional.empty();
 
         return provider;
     }
@@ -406,6 +408,28 @@ public class JWTAuthContextInfoProvider {
     @ConfigProperty(name = "smallrye.jwt.required.claims")
     Optional<Set<String>> requiredClaims;
 
+    /**
+     * TLS Trusted Certificate Path.
+     */
+    @Inject
+    @ConfigProperty(name = "smallrye.jwt.tls.certificate.path")
+    private Optional<String> tlsCertificatePath;
+
+    /**
+     * TLS Trust All.
+     * If this property is set to 'true' then HTTPS HostnameVerifier will trust all the hostnames.
+     */
+    @Inject
+    @ConfigProperty(name = "smallrye.jwt.tls.trust-all", defaultValue = "false")
+    private boolean tlsTrustAll;
+
+    /**
+     * TLS Trusted Hosts. Set this property if `smallrye.jwt.tls.trust-all` property is disabled.
+     */
+    @Inject
+    @ConfigProperty(name = "smallrye.jwt.tls.hosts")
+    private Optional<Set<String>> tlsTrustedHosts;
+
     @Produces
     Optional<JWTAuthContextInfo> getOptionalContextInfo() {
         String resolvedVerifyKeyLocation = !NONE.equals(verifyKeyLocation)
@@ -489,6 +513,9 @@ public class JWTAuthContextInfoProvider {
         contextInfo.setDefaultSubjectClaim(defaultSubClaim.orElse(null));
         SmallryeJwtUtils.setContextSubPath(contextInfo, subPath);
         contextInfo.setDefaultGroupsClaim(defaultGroupsClaim.orElse(null));
+        contextInfo.setTlsCertificatePath(tlsCertificatePath.orElse(null));
+        contextInfo.setTlsTrustedHosts(tlsTrustedHosts.orElse(null));
+        contextInfo.setTlsTrustAll(tlsTrustAll);
         SmallryeJwtUtils.setContextGroupsPath(contextInfo, groupsPath);
         contextInfo.setExpGracePeriodSecs(expGracePeriodSecs);
         contextInfo.setMaxTimeToLiveSecs(maxTimeToLiveSecs.orElse(null));

--- a/implementation/jwt-auth/src/test/java/io/smallrye/jwt/auth/principal/KeyLocationResolverTest.java
+++ b/implementation/jwt-auth/src/test/java/io/smallrye/jwt/auth/principal/KeyLocationResolverTest.java
@@ -24,12 +24,16 @@ import static org.mockito.Mockito.doAnswer;
 import static org.mockito.Mockito.when;
 
 import java.nio.charset.StandardCharsets;
+import java.security.cert.X509Certificate;
 import java.security.interfaces.RSAPublicKey;
+import java.util.Arrays;
 import java.util.Collections;
+import java.util.HashSet;
 
 import javax.crypto.SecretKey;
 import javax.crypto.spec.SecretKeySpec;
 
+import org.jose4j.http.Get;
 import org.jose4j.jwk.HttpsJwks;
 import org.jose4j.jwk.JsonWebKey;
 import org.jose4j.jwk.OctetSequenceJsonWebKey;
@@ -61,6 +65,8 @@ public class KeyLocationResolverTest {
     @Mock
     HttpsJwks mockedHttpsJwks;
     @Mock
+    Get mockedGet;
+    @Mock
     UrlStreamResolver urlResolver;
 
     RSAPublicKey rsaKey;
@@ -89,10 +95,81 @@ public class KeyLocationResolverTest {
         contextInfo.setJwksRefreshInterval(10);
 
         KeyLocationResolver keyLocationResolver = new KeyLocationResolver(contextInfo) {
-            protected HttpsJwks initializeHttpsJwks(String loc) {
+            protected HttpsJwks getHttpsJwks(String loc) {
                 return mockedHttpsJwks;
             }
+
+            protected Get getHttpGet() {
+                return mockedGet;
+            }
         };
+        Mockito.verify(mockedGet, Mockito.never()).setTrustedCertificates(Mockito.any(X509Certificate.class));
+        Mockito.verify(mockedGet, Mockito.never()).setHostnameVerifier(Mockito.any());
+        Mockito.verify(mockedHttpsJwks).setSimpleHttpGet(mockedGet);
+
+        RsaJsonWebKey jwk = new RsaJsonWebKey(rsaKey);
+        jwk.setKeyId("1");
+        when(mockedHttpsJwks.getJsonWebKeys()).thenReturn(Collections.singletonList(jwk));
+        keyLocationResolver = Mockito.spy(keyLocationResolver);
+        when(signature.getHeaders()).thenReturn(headers);
+        when(headers.getStringHeaderValue(JsonWebKey.KEY_ID_PARAMETER)).thenReturn("1");
+
+        assertEquals(rsaKey, keyLocationResolver.resolveKey(signature, emptyList()));
+        assertNull(keyLocationResolver.key);
+    }
+
+    @Test
+    public void testLoadRsaKeyFromHttpsJwksWithCertPathAndTrustAll() throws Exception {
+        JWTAuthContextInfo contextInfo = new JWTAuthContextInfo("https://github.com/my_key.jwks", "issuer");
+        contextInfo.setTlsCertificatePath("publicCrt.pem");
+        contextInfo.setTlsTrustAll(true);
+        contextInfo.setJwksRefreshInterval(10);
+
+        KeyLocationResolver keyLocationResolver = new KeyLocationResolver(contextInfo) {
+            protected HttpsJwks getHttpsJwks(String loc) {
+                return mockedHttpsJwks;
+            }
+
+            protected Get getHttpGet() {
+                return mockedGet;
+            }
+        };
+        Mockito.verify(mockedGet).setTrustedCertificates(Mockito.any(X509Certificate.class));
+        Mockito.verify(mockedGet).setHostnameVerifier(Mockito.any(AbstractKeyLocationResolver.TrustAllHostnameVerifier.class));
+        Mockito.verify(mockedHttpsJwks).setSimpleHttpGet(mockedGet);
+
+        RsaJsonWebKey jwk = new RsaJsonWebKey(rsaKey);
+        jwk.setKeyId("1");
+        when(mockedHttpsJwks.getJsonWebKeys()).thenReturn(Collections.singletonList(jwk));
+        keyLocationResolver = Mockito.spy(keyLocationResolver);
+        when(signature.getHeaders()).thenReturn(headers);
+        when(headers.getStringHeaderValue(JsonWebKey.KEY_ID_PARAMETER)).thenReturn("1");
+
+        assertEquals(rsaKey, keyLocationResolver.resolveKey(signature, emptyList()));
+        assertNull(keyLocationResolver.key);
+    }
+
+    @Test
+    public void testLoadRsaKeyFromHttpsJwksWithCertPathAndTrustedHosts() throws Exception {
+        JWTAuthContextInfo contextInfo = new JWTAuthContextInfo("https://github.com/my_key.jwks", "issuer");
+        contextInfo.setTlsCertificatePath("publicCrt.pem");
+        contextInfo.setTlsTrustedHosts(new HashSet<>(Arrays.asList("trusted-host")));
+        contextInfo.setJwksRefreshInterval(10);
+
+        KeyLocationResolver keyLocationResolver = new KeyLocationResolver(contextInfo) {
+            protected HttpsJwks getHttpsJwks(String loc) {
+                return mockedHttpsJwks;
+            }
+
+            protected Get getHttpGet() {
+                return mockedGet;
+            }
+        };
+        Mockito.verify(mockedGet).setTrustedCertificates(Mockito.any(X509Certificate.class));
+        Mockito.verify(mockedGet)
+                .setHostnameVerifier(Mockito.any(AbstractKeyLocationResolver.TrustedHostsHostnameVerifier.class));
+        Mockito.verify(mockedHttpsJwks).setSimpleHttpGet(mockedGet);
+
         RsaJsonWebKey jwk = new RsaJsonWebKey(rsaKey);
         jwk.setKeyId("1");
         when(mockedHttpsJwks.getJsonWebKeys()).thenReturn(Collections.singletonList(jwk));

